DataFormats/Provenance/src/BranchType.cc

```diff
@@ -16,52 +19,40 @@ namespace edm {
     std::string const branchEntryInfo = "BranchEntryInfo";  // backward compatibility
     std::string const productProvenance = "ProductProvenance";
 
-    // Prefixes
-    std::string const run = "Run";
-    std::string const lumi = "LuminosityBlock";
-    std::string const event = "Event";
+    std::array<std::string, NumBranchTypes> const branchTypeNames{{"Event", "LuminosityBlock", "Run", "ProcessBlock"}};
+    std::size_t constexpr eventLumiRunSize = 3;
+    using NameArray = const std::array<std::string, eventLumiRunSize>;
 
-    // Trees, branches, indices
-    std::string const runs = run + 's';
-    std::string const lumis = lumi + 's';
-    std::string const events = event + 's';
+    NameArray treeNames{{branchTypeNames[InEvent] + 's', branchTypeNames[InLumi] + 's', branchTypeNames[InRun] + 's'}};
```

Comment:
This assumes `InEvent==0`, `InLumi == 1` and `InRun` ==2 else the arrays get out of order. I think it would be better to just stick with the numbers and not the symbols.

Reply:
Hi Chris, I disagree with you on this one, because BranchType is in the interface of the functions that appear later in the file. It is the argument type to those functions. Those functions use BranchType to access the element in the std::array's this code builds. The dependence exists even if we build the arrays without the enum values.  Also, I don't like putting in magic numbers (0,1,2) here.

I was trying to think of something that could be done better here. Note that in the initial pull request, I simply extended the pre-existing coding style to include ProcessBlock. Then I switched to using std::array because Matti asked (and it's better that way). I could move branchTypeNames to the FWCore/Utilities/interface/BranchType.h file. Or I could add a stronger comment in BranchType.h about the fact that multiple things depend on the enum values (there already is such a comment).  Note that it's not only here, there are also places in IOPool that depend on the enum values.

Another alternative is to rewrite this file entirely in some different style in a different PR.

There are a lot of ways this could be done, but what is in the PR now will work.

Reply:
How about something along the following?
```cpp
std::array<std::string, eventLumiRunSize> makeNameArray(std::string_view postfix) {
  std::array<std::string, eventLumiRunSize> ret;
  for (auto i=0U; i != eventLumiRunSize; ++i) {
    ret[i] = branchTypeNames[i] + postfix;
  }
  return ret;
}
NameArray treeNames = makeNameArray("s");
NameArray metaTreeNames = makeNameArray(metaData);
```

Reply:
I'm with @makortel on this (as we require a runtime behavior anyway). The problem with using `InEvent`, `InLumi` and `InRun` is it makes the initialization appear invariant to the changes of value for the `In*` variables. However, that is not the case. If `InEvent` and `InLumi` values were swapped, the order of elements in the array `treeNames` would remain the same (i.e. `"Events"` would be in the 0th element) but using `treeNames[InEvent]` would return the string`"LuminosityBlocks"`.

Reply:
I added static_assert's requiring the enum values have the values that the design in this file requires. I also put in a makeNameArray function to reduce the code duplication. I pushed the commit so you can look and see if you like it. I'll not start the tests yet because a few more commits are coming tomorrow.